### PR TITLE
Use Dune for building

### PR DESCRIPTION
**Pull request by nojb**
This PR switches the build system from `make` to Dune (NB the configure script is still used).

The main advantage is to allow to compose `ocurl` easily in larger Dune monorepos, but it also yields a nice simplification of the build system (see the diff).

- `./configure` does not need to generate `Makefile`s anymore
- adapted the `release` target to use [`dune-release`](https://github.com/ocamllabs/dune-release) for easy releases to Github and the OPAM repository. (This required switching  `CHANGES` to Markdown format so that `dune-release` extracts the version from that file.)
- added a Github CI action using https://github.com/ocaml/setup-ocaml which tests ocurl on every version of OCaml from 4.03 to 4.12 on Linux and 4.12 on macOS (see a sample log at https://github.com/nojb/ocurl/actions/runs/1197150147)

### CODE REVIEW

File: dune-project [hunk 1]
```diff
@@ -0,0 +1,7 @@
+(lang dune 2.8)
```
[the remark this answers is not in the document]
Comment (ygrek) [reply dune-project]: better compatibiliy, not forcing users to rebuild everything

**Comment by ygrek** [comment]
I would on the contrary think that findlib name is more important compatibility-wise..

> Dune enforces that when a project consists of a single package they should have the same name. Here "project" is the thing that you use when you write opam install <project name> and "package" refers to the notion used by Findlib.

is this dune property or dune-release? I may side-step this issue entirely by not generating opam package with dune?

**Comment by nojb** [comment]
> I would on the contrary think that findlib name is more important compatibility-wise..

OK. Note that with the PR in its current form you keep both `ocurl` OPAM name **and** `curl` Findlib name (via the compatibility `META` shim), *and* `ocurl` can also be used as Findlib name. The alternative is to change the OPAM name from `ocurl` to `curl`, in which case no compatibility shim is needed.

> > Dune enforces that when a project consists of a single package they should have the same name. Here "project" is the thing that you use when you write opam install  and "package" refers to the notion used by Findlib.
> 
> is this dune property or dune-release? I may side-step this issue entirely by not generating opam package with dune?

I believe this is a `dune` property.

**Comment by ygrek** [comment]
Another argument is that the toplevel module name is Curl, so better keep library name `curl`. Made the change accordingly (make ocurl.opam a transitional package depending on curl.opam), tweaked ocamlc flags to my liking, restored back CHANGES.txt (rendered markdown hides the difference between `+` (additions) and others), and it is good to go.

Thanks a lot, that was a good piece of work.

Cherry-picked as 4eb931452400b857e12817faac7b9e65fffc1a42

**Comment by nojb** [comment]
> Cherry-picked as [4eb9314](https://github.com/ygrek/ocurl/commit/4eb931452400b857e12817faac7b9e65fffc1a42)

Thanks!

**Pull request closed without merging**